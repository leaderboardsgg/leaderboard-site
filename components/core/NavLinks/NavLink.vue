<template>
  <BaseButtonLink
    class="nav-link"
    v-bind="$attrs"
    v-on="$listeners"
  >
    <slot />
  </BaseButtonLink>
</template>

<style scoped>
  @layer components {
    .nav-link {
      @apply w-full mt-2 lg:mt-0 lg:mx-1 py-1 hover:bg-gray-100;
    }
  }
</style>
