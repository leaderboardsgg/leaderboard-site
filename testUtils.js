import Vue from 'vue';
import { RouterLinkStub } from '@vue/test-utils';
import userEvent from '@testing-library/user-event';
import { render } from '@testing-library/vue';

export const NuxtLinkStub = Vue.component('NuxtLinkStub', {
  extends: RouterLinkStub,
  render(h) {
    const temp = h(
      this.tag || 'a',
      {
        attrs: {
          ...this.$vnode.data.attrs,
          ...this.$attrs,
          href: this.to,
        },
      },
      this.$slots.default,
    );

    return temp;
  },
});

export const stubbedRender = (Component, options) => {
  return render(Component, {
    stubs: { NuxtLink: NuxtLinkStub, ...options?.stubs },
    ...options,
  });
};

export * from '@testing-library/vue';
export * from '@vue/test-utils';
export { userEvent as fireEvent };
