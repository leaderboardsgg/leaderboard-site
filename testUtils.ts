import Vue, { ComponentOptions } from 'vue';
import { RouterLinkStub, VueClass } from '@vue/test-utils';
import userEvent from '@testing-library/user-event';
import { render, RenderOptions } from '@testing-library/vue';

export const NuxtLinkStub = Vue.component<
  { tag: string; to: string },
  unknown,
  unknown
>('NuxtLinkStub', {
  extends: RouterLinkStub,
  render(h) {
    const temp = h(
      this.tag || 'a',
      {
        attrs: {
          ...this.$vnode.data?.attrs,
          ...this.$attrs,
          href: this.to,
        },
      },
      this.$slots.default,
    );

    return temp;
  },
});

export const stubbedRender = (
  Component: VueClass<Vue> | ComponentOptions<Vue>,
  options?: RenderOptions<Vue>,
) => {
  return render(Component, {
    stubs: { NuxtLink: NuxtLinkStub, ...options?.stubs },
    ...options,
  });
};

export * from '@testing-library/vue';
export * from '@vue/test-utils';
export { userEvent as fireEvent };
