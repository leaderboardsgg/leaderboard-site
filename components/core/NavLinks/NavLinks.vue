<template>
  <nav class="nav-link-container">
    <NavLink to="#">Games</NavLink>
    <NavLink to="#">About</NavLink>
  </nav>
</template>

<script lang="ts">
import Vue from 'vue'
import NavLink from './NavLink.vue';

export default Vue.extend({
  components: {
    NavLink
  }  
})
</script>

<style scoped>
  @layer components {
    .nav-link-container {
      @apply flex flex-col lg:flex-row items-center w-auto space-y-2 lg:space-y-0 text-center text-black;
    }
  }
</style>
