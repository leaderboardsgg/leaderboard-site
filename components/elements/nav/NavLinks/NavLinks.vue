<script setup lang="ts">
import NavLink from './NavLink/NavLink.vue'

interface NavLinksProps {
  navLinks?: {
    name: string
    to: string
  }[]
}

withDefaults(defineProps<NavLinksProps>(), {
  navLinks: () => [
    {
      name: 'Games',
      to: '#',
    },
    {
      name: 'About',
      to: '#',
    },
  ],
})
</script>

<template>
  <nav class="nav-link-container">
    <NavLink v-for="navLink in navLinks" v-bind="navLink" :key="navLink.name">
      {{ navLink.name }}
    </NavLink>
  </nav>
</template>

<style lang="postcss" scoped>
.nav-link-container {
  @apply flex flex-col lg:flex-row items-center w-auto space-y-2 lg:space-y-0 text-center text-black;
}
</style>
