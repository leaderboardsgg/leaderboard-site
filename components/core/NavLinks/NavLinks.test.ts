import { stubbedRender } from '@/testUtils';
import NavLinks from './NavLinks.vue';

describe('<NavLinks />', () => {
  const defaultProps = {
    navLinks: [
      { name: 'Home', to: '/' },
      { name: 'About', to: '/about' },
      { name: 'Contact', to: '/contact' },
    ],
  };

  it('should render without crashing', () => {
    const { unmount } = stubbedRender(NavLinks, {
      props: defaultProps,
    });

    unmount();
  });

  it('renders correctly', () => {
    const { container } = stubbedRender(NavLinks, {
      props: defaultProps,
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the same amount of <NavLink /> components as there are items in the navLinks props', () => {
    const { container, getByText } = stubbedRender(NavLinks, {
      props: defaultProps,
    });

    expect(container.firstChild?.childNodes.length).toEqual(
      defaultProps.navLinks.length,
    );
    defaultProps.navLinks.forEach((navLink) => {
      expect(getByText(navLink.name)).toBeInTheDocument();
    });
  });
});
