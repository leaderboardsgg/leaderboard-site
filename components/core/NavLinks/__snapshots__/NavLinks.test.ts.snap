// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<NavLinks /> renders correctly 1`] = `
<nav
  class="nav-link-container"
>
  <a
    class="button-link nav-link"
    href="/"
    name="Home"
    to="/"
  >
    
    Home
  
  </a>
  <a
    class="button-link nav-link"
    href="/about"
    name="About"
    to="/about"
  >
    
    About
  
  </a>
  <a
    class="button-link nav-link"
    href="/contact"
    name="Contact"
    to="/contact"
  >
    
    Contact
  
  </a>
</nav>
`;
